src/librustdoc/html/markdown.rs

```diff
@@ -1057,7 +1057,14 @@ fn markdown_summary_with_limit(md: &str, length_limit: usize) -> (String, bool)
         *text_length += text.len();
     };
 
-    'outer: for event in Parser::new_ext(md, Options::ENABLE_STRIKETHROUGH) {
+    // NOTE: Make sure to update the same variable in `plain_text_summary`
+    //       if/when you update this one. They have to be duplicated because of a typesystem thing.
+    let mut broken_link_callback =
+        |broken_link: BrokenLink<'_>| Some(("#".into(), broken_link.reference.to_owned().into()));
```

Comment:
This treats *every* broken link as valid, right? Can we instead use the same logic as for intra-doc links and only replace it if it was resolved?

Reply:
Well we _could_, but we’d have to thread the intra-doc link information through somehow. If you would like me to do that, could you give me some instructions?

Reply:
This info is available from `item.attrs.links`: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/types/struct.Attributes.html#structfield.links. It looks like that's not currently threaded through to `short_markdown_summary`, but everywhere that calls `summary` calls it on an `Item`. I'd suggest making this a method on `Item` instead and calling `item.short_markdown_summary()` instead, which gives you access to all the info on the item.

See https://github.com/rust-lang/rust/blob/5019791e2d4ac67b4c5c56853600a02f4b84e515/src/librustdoc/html/markdown.rs#L365 for an example of actually using the link - I don't expect this to be as complicated since it's just stripping the link altogether, and only for `[]` style links.

Reply:
Hmm, will I then I have to duplicate `broken_link_callback` as a closure after all?

Reply:
I don't see why? You're stripping the links in both cases, right?

Reply:
Yes, but what would we pass as the callback?

Reply:
The same closure as now, but extracted into a function.

Can you explain why this would be any different with a function taking `item` than it is now? I'm really confused.

Reply:
> Can you explain why this would be any different with a function taking `item` than it is now? I'm really confused.

Sure! It's possible I'm the one who's actually confused :)

Basically, you're suggesting that the `markdown_summary_with_limit` and `plain_text_summary` functions take an additional parameter (`self: &Item`) that is used to check what links are intra-doc links. But, that information has to be available in `summary_broken_link_callback` so that we only fake-resolve (i.e. use `#` as the link) intra-doc links. This would be easy in a closure: we would just capture the argument passed into `markdown_summary_with_limit`/`plain_text_summary`. But how do we do that with a shared `fn`? Do we "curry" the callback function and write it as something like below?

```rust
fn summary_broken_link_callback<'a>(item: &Item) -> impl FnMut(BrokenLink<'_>) -> (CowStr<'a>, CowStr<'a>) {
    |broken_link: BrokenLink<'_>| /* ... */ item.attrs.links /* ... */
}
```

Was that clearer?

Reply:
Got it, thanks, I forgot `Parser` doesn't allow passing arguments into the callback.

I think we should wait to think about reducing duplication until we see how much code it actually is. If it's only ~3 lines it's probably better to repeat it, if it's more I have a few not fully-fledged out ideas.

Reply:
Yeah, I'll just add back the `NOTE:` comment so that we keep them in sync.

Reply:
It's been a while since we lasted discussed this. It looks like the last thing we talked about is letting their be a little duplication and moving the summary functions to be on `Item`?

Reply:
It's coming back to me now: The reason I temporarily abandoned this is because I was having trouble with the lifetimes of the callback for the summary functions not on `Item`. I think it might have been the dreaded higher-ranked subtype error. I'll post the error if/when I get it.

Reply:
So I tried this type signature:

```rust
fn markdown_summary_with_limit(
    md: &str,
    length_limit: usize,
    broken_link_callback: Option<F>,
) -> (String, bool)
where
    F: for<'a> FnMut(BrokenLink<'_>) -> Option<(CowStr<'a>, CowStr<'a>)>,
{ /* ... */ }
```

but that gives a bunch of errors like:

```
error[E0582]: binding for associated type `Output` references lifetime `'a`, which does not appear in the trait input types
    --> src/librustdoc/html/markdown.rs:1030:41
     |
1030 |     F: for<'a> FnMut(BrokenLink<'_>) -> Option<(CowStr<'a>, CowStr<'a>)>,
     |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

What are the correct lifetimes? I feel like this needs higher-ranked lifetimes (which I added) but as you can see it doesn't work. I need to communicate to the compiler that the return type lifetimes are totally unrelated from the input lifetimes.

Reply:
@jyn514 friendly ping :)

If you need to focus on other stuff, don't worry about looking at this now, but you seemed eager to get this working.

Reply:
Try `F: FnMut(BrokenLink<'_>) -> Option<(CowStr<'static>, CowStr<'static>)>`. That has the same meaning without introducing a new lifetime.